### PR TITLE
SALTO-2643 Okta adapter setup

### PR DESCRIPTION
**Pull request by shir-reifenberg**
Initial setup of Okta adapter with fetch and partial deploy

---

_Additional context for reviewer_

Add Okta as a new adapter using the infrastructure and [okta's swagger file](https://raw.githubusercontent.com/okta/okta-management-openapi-spec/master/dist/spec.yaml)

---

Release Notes:

Okta adapter:

- Add support in fetch
- Partial deploy of `Group`, `Applicaiton` and `AppUser` instances

---

### CODE REVIEW

File: packages/okta-adapter/src/filters/standard_roles.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+import { Element, InstanceElement } from '@salto-io/adapter-api'
+import { elements as elementUtils } from '@salto-io/adapter-components'
+import { naclCase } from '@salto-io/adapter-utils'
+import { FilterCreator } from '../filter'
+import { findObject } from '../utils'
+import { ROLE_TYPE_NAME, OKTA } from '../constants'
+
+const { RECORDS_PATH } = elementUtils
+
+const ROLE_TYPE_TO_LABEL: Record<string, string> = {
```
Comment (netama) [inline packages/okta-adapter/src/filters/standard_roles.ts:25]: nit: if we have a link to the docs can you please add it here for reference?

File: packages/okta-adapter/system_config_doc.md [hunk 2]
```diff
@@ -0,0 +1,9 @@
+# Okta system configuration
+## Default Configuration
+TODO
```
Comment (netama) [inline packages/okta-adapter/system_config_doc.md:3]: let's update when the config stabilizes

**Review by netama — COMMENTED** [review]
will continue later

File: packages/okta-adapter/src/client/pagination.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+import { client as clientUtils } from '@salto-io/adapter-components'
+import _ from 'lodash'
+import * as parse from 'parse-link-header'
```
Comment (netama) [inline packages/okta-adapter/src/client/pagination.ts:18]: cool :)

File: packages/okta-adapter/src/config.ts [hunk 4]
```diff
@@ -0,0 +1,712 @@
...
+export const API_DEFINITIONS_CONFIG = 'apiDefinitions'
+
+export type OktaClientConfig = clientUtils.ClientBaseConfig<clientUtils.ClientRateLimitConfig>
+
+export type OktaFetchConfig = configUtils.UserFetchConfig
+export type OktaApiConfig = configUtils.AdapterSwaggerApiConfig & {
+  settingsSwagger?: {
+    typeNameOverrides?: configUtils.TypeNameOverrideConfig[]
+  }
+}
```
Comment (netama) [inline packages/okta-adapter/src/config.ts:36]: can remove (we don't have a separate `settingsSwagger`, that probably came from zuora)
```suggestion
export type OktaApiConfig = configUtils.AdapterSwaggerApiConfig
```

File: packages/okta-adapter/src/config.ts [hunk 5]
```diff
@@ -0,0 +1,712 @@
...
+
+export type OktaConfig = {
+  [CLIENT_CONFIG]?: OktaClientConfig
+  [FETCH_CONFIG]: OktaFetchConfig
+  [API_DEFINITIONS_CONFIG]: OktaApiConfig
+}
+
+const DEFAULT_ID_FIELDS = ['name']
+export const FIELDS_TO_OMIT: configUtils.FieldToOmitType[] = [
+  // TODO
```
Comment (netama) [inline packages/okta-adapter/src/config.ts:46]: are you still planning to add anything here? (links maybe? if so let's add a ticket for when we can add that and link it from here, if not then let's just remove the TODO)

File: packages/okta-adapter/src/config.ts [hunk 6]
```diff
@@ -0,0 +1,712 @@
...
+    transformation: {
+      fieldTypeOverrides: [
+        { fieldName: 'appUsers', fieldType: 'list<AppUser>' },
+        { fieldName: 'CSRs', fieldType: 'list<Csr>' },
+        { fieldName: 'assignedGroups', fieldType: 'list<ApplicationGroupAssignment>' },
+        { fieldName: 'profileEnrollment', fieldType: 'string' },
+        { fieldName: 'accessPolicy', fieldType: 'string' },
+      ],
+      standaloneFields: [{ fieldName: 'appUsers' }],
+      // TODO SALTO-2644
```
Comment (netama) [inline packages/okta-adapter/src/config.ts:165]: maybe add a short explanation about what the issue is inside the comment?

File: packages/okta-adapter/src/config.ts [hunk 7]
```diff
@@ -0,0 +1,712 @@
...
+        { fieldName: 'CSRs', fieldType: 'list<Csr>' },
+      ],
+    },
+  },
+  api__v1__features: {
+    request: {
+      url: '/api/v1/features',
+      recurseInto: [
+        {
+          // Features that needs to be enabled in order to enable the feature
```
Comment (netama) [inline packages/okta-adapter/src/config.ts:291]: nit:
```suggestion
          // Additional features that need to be enabled in order to enable the feature
```
(if I understood correctly)

File: packages/okta-adapter/src/config.ts [hunk 8]
```diff
@@ -0,0 +1,712 @@
...
+          type: 'api__v1__features___featureId___dependencies@uuuuuu_00123_00125uu',
+          toField: 'featureDependencies',
+          context: [{ name: 'featureId', fromField: 'id' }],
+        },
+        {
+          // Features that needs to be disabled in order to enable the feature
```
Comment (netama) [inline packages/okta-adapter/src/config.ts:297]: not sure I understand this one - why do they need to be disabled in order to enable this? should it have been `disabled`? (and if so - do we get anything from surfacing dependents here that we don't get by surfacing only the "dependencies" direction in all features?)

File: packages/okta-adapter/src/config.ts [hunk 9]
```diff
@@ -0,0 +1,712 @@
...
+    transformation: {
+      fieldTypeOverrides: [
+        { fieldName: 'featureDependencies', fieldType: 'list<Feature>' },
+        { fieldName: 'featureDependents', fieldType: 'list<Feature>' },
+      ],
+    },
+  },
+  AuthenticatorEnrollmentPolicies: {
+    request: {
+      url: '/api/v1/policies',
```
Comment (netama) [inline packages/okta-adapter/src/config.ts:315]: maybe add a comment before all types we get from the `'/api/v1/policies'` url, explaining that they are split across different types and where the full list of types can be found?

File: packages/okta-adapter/src/config.ts [hunk 10]
```diff
@@ -0,0 +1,712 @@
...
+      recurseInto: [
+        {
+          type: 'api__v1__policies___policyId___rules@uuuuuu_00123_00125uu',
+          toField: 'policyRules',
+          context: [{ name: 'policyId', fromField: 'id' }],
+        },
+      ],
+    },
+  },
+  // TODO returns 400 bad request
```
Comment (netama) [inline packages/okta-adapter/src/config.ts:403]: let's add a ticket?

File: packages/okta-adapter/src/config.ts [hunk 11]
```diff
@@ -0,0 +1,712 @@
...
+  url: 'https://raw.githubusercontent.com/okta/okta-management-openapi-spec/master/dist/spec.yaml',
+  additionalTypes: [
+    { typeName: 'AuthenticatorEnrollmentPolicies', cloneFrom: 'api__v1__policies' },
+    { typeName: 'GlobalSessionPolicies', cloneFrom: 'api__v1__policies' },
+    { typeName: 'AuthenticationPolicies', cloneFrom: 'api__v1__policies' },
+    { typeName: 'ProfileEnrollmentPolicies', cloneFrom: 'api__v1__policies' },
+    { typeName: 'IdentityProviderRoutingRules', cloneFrom: 'api__v1__policies' },
+    { typeName: 'PasswordPolicies', cloneFrom: 'api__v1__policies' },
+    { typeName: 'OAuthAuthorizationPolicies', cloneFrom: 'api__v1__policies' },
+    // TODO this is not the right type to clone from, another solution is needed
```
Comment (netama) [inline packages/okta-adapter/src/config.ts:609]: let's add a ticket?

File: packages/okta-adapter/src/config.ts [hunk 12]
```diff
@@ -0,0 +1,712 @@
...
+    { typeName: 'RolePage', cloneFrom: 'api__v1__groups___groupId___roles@uuuuuu_00123_00125uu' },
+  ],
+}
+
+/* eslint-disable max-len */
```
Comment (netama) [inline packages/okta-adapter/src/config.ts:614]: let's remove this comment and fix the relevant lines (or disable only for those specific lines)

File: packages/okta-adapter/src/config.ts [hunk 13]
```diff
@@ -0,0 +1,712 @@
...
+}
+
+
+export const DEFAULT_API_DEFINITIONS: OktaApiConfig = {
+  swagger: DEFAULT_SWAGGER_CONFIG,
+  typeDefaults: {
+    transformation: {
+      idFields: DEFAULT_ID_FIELDS,
+      fieldsToOmit: FIELDS_TO_OMIT,
+      // TODO: currently these fields can be nested
```
Comment (netama) [inline packages/okta-adapter/src/config.ts:669]: if you think we should eventually do this let's add a ticket with details about if/when  (I tend to think if this is risky we should just have a constant and reuse it in the types where it can be added, but not use it in the default config - like we do in [zendesk](https://github.com/salto-io/salto/blob/main/packages/zendesk-adapter/src/config.ts#L78))

File: packages/okta-adapter/src/config.ts [hunk 14]
```diff
@@ -0,0 +1,712 @@
...
+  NetworkZone: ['api__v1__zones'],
+  Domain: ['DomainListResponse'],
+  Role: ['RolePage'],
+}
+
+
+export const DEFAULT_API_DEFINITIONS: OktaApiConfig = {
+  swagger: DEFAULT_SWAGGER_CONFIG,
+  typeDefaults: {
+    transformation: {
```
Comment (netama) [inline packages/okta-adapter/src/config.ts:666]: I think this is a good time to add `serviceIdField`s (I'm assuming the default will be `id`? but let's adjust it if there are other types). we can do it in a separate PR if you prefer - but then let's have a ticket for it

File: packages/okta-adapter/src/filters/app_deployment.ts [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+import { Change, InstanceElement, isInstanceChange, getChangeData, isAdditionOrModificationChange, isModificationChange, ModificationChange, isAdditionChange } from '@salto-io/adapter-api'
+import { config as configUtils, elements as elementUtils, deployment } from '@salto-io/adapter-components'
+import { APPLICATION_TYPE_NAME } from '../constants'
+import OktaClient from '../client/client'
+import { OktaConfig, API_DEFINITIONS_CONFIG } from '../config'
+import { FilterCreator } from '../filter'
+import { deployChanges, defaultDeployChange, deployEdges } from '../deployment'
+
+const INACTIVE_STATUS = 'INACTIVE'
+const APP_ASSIGNMENT_FIELDS: Record<string, configUtils.DeploymentRequestsByAction> = {
```
Comment (netama) [inline packages/okta-adapter/src/filters/app_deployment.ts:26]: why can't we add this as part of the main config? is it because of the parameter?
if so - can you please create a ticket for what's missing?

File: packages/okta-adapter/src/filters/app_deployment.ts [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+    fieldsToIgnore,
+    isAdditionChange(change) && getChangeData(change).value.status === 'INACTIVE' ? { activate: 'false' } : undefined
+  )
+
+  if (isAdditionOrModificationChange(change)) {
+    await deployEdges(change, APP_ASSIGNMENT_FIELDS, client)
+  }
+}
+
+const filterCreator: FilterCreator = ({ client, config }) => ({
```
Comment (netama) [inline packages/okta-adapter/src/filters/app_deployment.ts:101]: let's add a docstring (if there's a simple way to explain why this is not deployed using the default deploy)?

File: packages/okta-adapter/src/filters/referenced_id_fields.ts [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+* limitations under the License.
+*/
+import { filters } from '@salto-io/adapter-components'
+import { FilterContext } from '../config'
+import { FilterCreator, FilterResult } from '../filter'
+import OktaClient from '../client/client'
+
+/**
+ *  Resolves references in elements name using referenced idFields
+ *
```
Comment (netama) [inline packages/okta-adapter/src/filters/referenced_id_fields.ts:23]: nit: empty line can be removed

File: packages/okta-adapter/src/filters/replace_object_with_id.ts [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+            return value.type !== undefined ? value.type : value
+          }
+          return value.id !== undefined ? value.id : value
+        })
+    }
+  })
+}
+
+/**
+ * Some instances have references that contains the entire referenced object
```
Comment (netama) [inline packages/okta-adapter/src/filters/replace_object_with_id.ts:45]: nit: `contains` -> `contain`

File: packages/okta-adapter/src/filters/standard_roles.ts [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+  MOBILE_ADMIN: 'Mobile Administrator',
+  ORG_ADMIN: 'Organizational Administrator',
+  READ_ONLY_ADMIN: 'Read-Only Administrator',
+  REPORT_ADMIN: 'Report Administrator',
+  SUPER_ADMIN: 'Super Administrator',
+  USER_ADMIN: 'Group Administrator',
+}
+
+/**
+ * Adds standard role instances
```
Comment (netama) [inline packages/okta-adapter/src/filters/standard_roles.ts:39]: nit: maybe add something like `, which are not returned by the API but are useful for impact analysis`

File: packages/okta-adapter/src/reference_mapping.ts [hunk 20]
```diff
@@ -0,0 +1,139 @@
...
+import { APPLICATION_TYPE_NAME, GROUP_TYPE_NAME, USER_TYPE_NAME, IDENTITY_PROVIDER_TYPE_NAME, USERTYPE_TYPE_NAME, FEATURE_TYPE_NAME, POLICY_TYPE_NAME, NETWORK_ZONE_TYPE_NAME, APP_USER_TYPE_NAME, ROLE_TYPE_NAME } from './constants'
+
+
+export const referencesRules: referenceUtils.FieldReferenceDefinition<never>[] = [
+  {
+    src: { field: 'assignedGroups', parentTypes: [APPLICATION_TYPE_NAME] },
+    serializationStrategy: 'id',
+    target: { type: GROUP_TYPE_NAME },
+  },
+  // TODO move the reference to user to parent annotation?
```
Comment (netama) [inline packages/okta-adapter/src/reference_mapping.ts:28]: let's add a ticket with more context?

File: packages/okta-adapter/src/utils.ts [hunk 21]
```diff
@@ -0,0 +1,62 @@
+/*
```
Comment (netama) [inline packages/okta-adapter/src/utils.ts:1]: this entire file was copied from jira, right? if this is useful let's just add it under adapter-components and remove the duplicates.
specifically for `findObject`, I think `findObjectType` in `adapter-utils` might do the trick, unless @alonstern wanted the optimization of not looking at the adapter name?

**Review by netama — COMMENTED** [review]
looks great 🏆  added some comments, mostly around cleaning up some leftovers from the initial setup (some are my fault 🙃 )

File: packages/okta-adapter/src/config.ts [hunk 22]
```diff
@@ -0,0 +1,712 @@
...
+    },
+    [API_DEFINITIONS_CONFIG]: {
+      refType: createSwaggerAdapterApiConfigType({
+        adapter: OKTA,
+      }),
+    },
+  },
+  annotations: {
+    [CORE_ANNOTATIONS.DEFAULT]: _.omit(DEFAULT_CONFIG, API_DEFINITIONS_CONFIG),
+  },
```
Comment (netama) [inline packages/okta-adapter/src/config.ts:706]: we should also add this now so we can enforce there are only supported configuration properties (this is a new feature that was added after you started the adapter) - here and anywhere else we create a config object type (I think this is the only one for now):
```suggestion
    [CORE_ANNOTATIONS.ADDITIONAL_PROPERTIES]: false,
  },
```

File: packages/okta-adapter/src/config.ts (hunk 14 above)
Comment (shir-reifenberg) [reply packages/okta-adapter/src/config.ts:666]: Yeah from what I saw all types have `id` field so I'll add it

File: packages/okta-adapter/src/filters/app_deployment.ts (hunk 15 above)
Comment (shir-reifenberg) [reply packages/okta-adapter/src/filters/app_deployment.ts:26]: Because the 'typeName' is not really a type name but the relevant field's name

File: packages/okta-adapter/src/auth.ts [hunk 23]
```diff
@@ -0,0 +1,33 @@
...
+export type AccessTokenCredentials = {
+  baseUrl: string
+  token: string
+}
+
+export const accessTokenCredentialsType = createMatchingObjectType<AccessTokenCredentials>({
+  elemID: new ElemID(constants.OKTA),
+  fields: {
+    baseUrl: { refType: BuiltinTypes.STRING, annotations: { _required: true } },
+    token: { refType: BuiltinTypes.STRING, annotations: { _required: true } },
```
[the remark this answers is not in the document]
Comment (shir-reifenberg) [reply packages/okta-adapter/src/auth.ts:29]: I added to the baseUrl do you think we should add to the token as well?

[the remark this answers is not in the document]
Comment (netama) [reply packages/okta-adapter/src/auth.ts:29]: no I think that's ok if there are guidelines in the README, sorry I put this on the wrong line :)

File: packages/okta-adapter/src/config.ts (hunk 5 above)
Comment (shir-reifenberg) [reply packages/okta-adapter/src/config.ts:46]: I'll remove it for now

File: packages/okta-adapter/src/utils.ts (hunk 21 above)
Comment (shir-reifenberg) [reply packages/okta-adapter/src/utils.ts:1]: Actually I just noticed that I don't use the other functions so I'll remove this file for now

File: packages/okta-adapter/src/config.ts (hunk 8 above)
Comment (shir-reifenberg) [reply packages/okta-adapter/src/config.ts:297]: You're right it's redundant I miss understood the api doc

File: packages/okta-adapter/src/config.ts (hunk 13 above)
Comment (shir-reifenberg) [reply packages/okta-adapter/src/config.ts:669]: Yap I agree, anyway there's a ticket for this

File: packages/okta-adapter/src/fetch_criteria.ts [hunk 24]
```diff
@@ -0,0 +1,26 @@
...
+import { elements as elementUtils } from '@salto-io/adapter-components'
+import { regex } from '@salto-io/lowerdash'
+
+const nameCriterion: elementUtils.query.QueryCriterion = ({
+  instance,
+  value,
+}): boolean => regex.isFullRegexMatch(instance.value.name, value)
+
+export default {
+  name: nameCriterion,
```
[the remark this answers is not in the document]
Comment (shir-reifenberg) [reply packages/okta-adapter/src/fetch_criteria.ts:25]: I remove it for now :)

File: packages/okta-adapter/src/deployment.ts [hunk 25]
```diff
@@ -0,0 +1,187 @@
...
+      const removeConfig = deployRequestByField[fieldName].remove
+      if (isDefined(fieldValuesToRemove) && isDefined(removeConfig)) {
+        await awu(fieldValuesToRemove).forEach(fieldValue =>
+          deployEdge({ source: instanceId, target: fieldValue }, removeConfig, fieldName))
+      }
+    }
+  })
+}
+
+export const deployChanges = async <T extends Change<ChangeDataType>>(
```
[the remark this answers is not in the document]
Comment (shir-reifenberg) [reply packages/okta-adapter/src/deployment.ts:167]: talked IRL and decided to do this in a different PR

File: packages/jira-adapter/src/deployment/standard_deployment.ts [hunk 26]
```diff
@@ -78,6 +78,7 @@ export const defaultDeployChange = async ({
     apiDefinitions.types[getChangeData(change).elemID.typeName]?.deployRequests,
     fieldsToIgnore,
     additionalUrlVars,
+    undefined,
```
Comment (shir-reifenberg) [inline packages/jira-adapter/src/deployment/standard_deployment.ts:81]: Temporary change until I merge the [deployChange params change](https://github.com/salto-io/salto/pull/3362)

**Review by netama — APPROVED** [review]
